Exercícios da aula 3 de pandas - Módulo 2

**Dados**: Realize os passos seguintes utilizando o mesmo dataset do íris da aula anterior ('https://archive.ics.uci.edu/ml/machine-learning-databases/iris/iris.data').



a. Use o pandas para ler o arquivo como um dataframe. Obs: precisa ler o dataframe sem que a primeira linha corresponda ao nome das colunas.

In [ ]:
import pandas as pd

In [ ]:
url = 'https://archive.ics.uci.edu/ml/machine-learning-databases/iris/iris.data'
data = pd.read_csv(url, header=None)
data.head()

0    1    2    3            4
0  5.1  3.5  1.4  0.2  Iris-setosa
1  4.9  3.0  1.4  0.2  Iris-setosa
2  4.7  3.2  1.3  0.2  Iris-setosa
3  4.6  3.1  1.5  0.2  Iris-setosa
4  5.0  3.6  1.4  0.2  Iris-setosa

----

**b**. Sabendo que as colunas correspondem, nessa ordem, a: <br>

  1. sepal length (cm)
  2. sepal width (cm)
  3. petal length (cm)
  4. petal width (cm)
  5. class: <br>
      - Iris Setosa <br>
      - Iris Versicolor <br>
      - Iris Virginica <br>

leia novamente o arquivo passando o nome das colunas como argumento.

In [ ]:
# para fazer uma comparação dessa vez vou chamar de tabela
tabela = pd.read_csv('https://archive.ics.uci.edu/ml/machine-learning-databases/iris/iris.data', sep=",", header=None, names = ["sepalLength", "sepalWidth", "petalLength","petalWidth", "classType"])
tabela

sepalLength  sepalWidth  petalLength  petalWidth       classType
0            5.1         3.5          1.4         0.2     Iris-setosa
1            4.9         3.0          1.4         0.2     Iris-setosa
2            4.7         3.2          1.3         0.2     Iris-setosa
3            4.6         3.1          1.5         0.2     Iris-setosa
4            5.0         3.6          1.4         0.2     Iris-setosa
..           ...         ...          ...         ...             ...
145          6.7         3.0          5.2         2.3  Iris-virginica
146          6.3         2.5          5.0         1.9  Iris-virginica
147          6.5         3.0          5.2         2.0  Iris-virginica
148          6.2         3.4          5.4         2.3  Iris-virginica
149          5.9         3.0          5.1         1.8  Iris-virginica

[150 rows x 5 columns]

Também poderia só adicionar somente o nome das colunas no df que já baixei 

In [ ]:
nomes = ['sepal_length', 'sepal_width', 'petal_length', 'petal_width', 'class_type']
data = pd.read_csv(url, names = nomes)
data.head()

sepal_length  sepal_width  petal_length  petal_width   class_type
0           5.1          3.5           1.4          0.2  Iris-setosa
1           4.9          3.0           1.4          0.2  Iris-setosa
2           4.7          3.2           1.3          0.2  Iris-setosa
3           4.6          3.1           1.5          0.2  Iris-setosa
4           5.0          3.6           1.4          0.2  Iris-setosa

----
**c**. Calcule a média de cada coluna numérica para cada um dos tipos de íris indicados na coluna "class" sem utilizar métodos que não foram ensinados na aula de hoje. Utilize o loop for para reduzir quantidade de linhas. <br>
Existe diferença entre elas?

In [ ]:
# Para seber quais são as colunas que temos, posso usar os comandos .columns()
tabela.columns

Index(['sepalLength', 'sepalWidth', 'petalLength', 'petalWidth', 'classType'], dtype='object')

In [ ]:
# Para saber quais são os tipos de classes, posso usar  o comendo .unique()
tabela['classType'].unique()

array(['Iris-setosa', 'Iris-versicolor', 'Iris-virginica'], dtype=object)

Para fazer o levantamento das médias, posso utilizar varios métodos 

1. Filtro simples
2. Filtro com query
3. Filtro simples dentro do loop for

1- usando um filtro simples


In [ ]:
Iris_Setosa = data[data['class_type'] == 'Iris-setosa'].mean()
print (Iris_Setosa)

Iris_Versicolor = data[data['class_type'] == 'Iris-versicolor'].mean() 
print (Iris_Versicolor)

Iris_Virginica = data[data['class_type'] == 'Iris-virginica'].mean()
print (Iris_Virginica)

sepal_length    5.006
sepal_width     3.418
petal_length    1.464
petal_width     0.244
dtype: float64
sepal_length    5.936
sepal_width     2.770
petal_length    4.260
petal_width     1.326
dtype: float64
sepal_length    6.588
sepal_width     2.974
petal_length    5.552
petal_width     2.026
dtype: float64


2- Usando o método query()

In [ ]:
Iris_Setosa = data.query(" class_type == 'Iris-setosa' " ).mean()
print (Iris_Setosa)

Iris_Versicolor = data.query(" class_type == 'Iris-versicolor' ").mean()
print (Iris_Versicolor)

Iris_Virginica = data.query(" class_type == 'Iris-virginica' ").mean()
print (Iris_Virginica)

sepal_length    5.006
sepal_width     3.418
petal_length    1.464
petal_width     0.244
dtype: float64
sepal_length    5.936
sepal_width     2.770
petal_length    4.260
petal_width     1.326
dtype: float64
sepal_length    6.588
sepal_width     2.974
petal_length    5.552
petal_width     2.026
dtype: float64


3- Usando o loop for 

Aqui a ideia é um for para ler a lista de colunas e um outro for para ler os tipos de de classes
E então utilizar as variáveis do for (i e j) para fazer o filtro 

In [ ]:
for i in ['sepalLength', 'sepalWidth', 'petalLength', 'petalWidth']:  # lendo a lista de colunas que quero fazer a média
  for j in ['Iris-setosa', 'Iris-versicolor', 'Iris-virginica']: # lendo a lista dos tipos de classes
    media = tabela[tabela['classType'] == j][i].mean()
    print(f'{i}: {j} - ' , media)

print ("-----")
# poderia ter feito o filtro dentro do print. ex: 
print(f'{i}: {j}' , tabela[tabela['classType']==j][i].mean())

sepalLength: Iris-setosa -  5.005999999999999
sepalLength: Iris-versicolor -  5.936
sepalLength: Iris-virginica -  6.587999999999998
sepalWidth: Iris-setosa -  3.4180000000000006
sepalWidth: Iris-versicolor -  2.7700000000000005
sepalWidth: Iris-virginica -  2.9739999999999998
petalLength: Iris-setosa -  1.464
petalLength: Iris-versicolor -  4.26
petalLength: Iris-virginica -  5.552
petalWidth: Iris-setosa -  0.2439999999999999
petalWidth: Iris-versicolor -  1.3259999999999998
petalWidth: Iris-virginica -  2.026
-----
petalWidth: Iris-virginica 2.026


Usando o loop for também posso criar um df com os valores da média 

In [ ]:
classes = ["Iris-setosa", "Iris-versicolor", "Iris-virginica"]
medias = pd.DataFrame()

for item in classes:
    media = tabela[tabela["classType"] == item].iloc[:, 0:4].mean(axis=0)
    medias[item] = media
    
medias

Iris-setosa  Iris-versicolor  Iris-virginica
sepalLength        5.006            5.936           6.588
sepalWidth         3.418            2.770           2.974
petalLength        1.464            4.260           5.552
petalWidth         0.244            1.326           2.026

Respondendo a parte final da questão: Sim existe uma direfença entre as médias dos tipos de classes

- Iris_Setosa 

        * sepal_length    5.006
        * sepal_width     3.418
        * petal_length    1.464
        * petal_width     0.244


-  Iris_Versicolor

        * sepal_length    5.936
        * sepal_width     2.770
        * petal_length    4.260
        * petal_width     1.326


- Iris_Virginica

        * sepal_length    6.588
        * sepal_width     2.974
        * petal_length    5.552
        * petal_width     2.026


---
**d.** Adicione uma única coluna com a média de 'sepal length' para cada um dos tipos de íris indicados na coluna "class". Utilize o loop for e faça isso sem utilizar métodos que não foram ensinados na aula de hoje.

Posso fazer isso de duas formas 
1. Utiliazando os valores que já havia calculado anteriromente, o que é facil quando se tem poucos dados
2. Utilizando um loop for e recalculando o dados e já adicionando em uma nova coluna 

1- fazendo manualmente com os dados calculados

In [ ]:
# crio a coluna e passo o valor calculado 
data.loc [data ['class_type'] == 'Iris-setosa', 'mean_sepal_length'] = 5.006
data.loc [data ['class_type'] == 'Iris-versicolor', 'mean_sepal_length'] = 5.936
data.loc [data ['class_type'] == 'Iris-virginica', 'mean_sepal_length'] = 6.588

In [ ]:
data

sepal_length  sepal_width  ...      class_type  mean_sepal_length
0             5.1          3.5  ...     Iris-setosa              5.006
1             4.9          3.0  ...     Iris-setosa              5.006
2             4.7          3.2  ...     Iris-setosa              5.006
3             4.6          3.1  ...     Iris-setosa              5.006
4             5.0          3.6  ...     Iris-setosa              5.006
..            ...          ...  ...             ...                ...
145           6.7          3.0  ...  Iris-virginica              6.588
146           6.3          2.5  ...  Iris-virginica              6.588
147           6.5          3.0  ...  Iris-virginica              6.588
148           6.2          3.4  ...  Iris-virginica              6.588
149           5.9          3.0  ...  Iris-virginica              6.588

[150 rows x 6 columns]

2- utilizando o loop for a automatizando essa processo

In [ ]:
'''
Para fazer o for o que vou precicar é, atribui à uma nova coluna o valor da média
nesse caso as estruturas dos códigos será muito parecida, pois para criara a coluna e fazer a média é basicamente a mesma coisa
1 - Criando a nova coluna :
  df.loc [df ['col base'] == 'o quero', 'nome da coluna nova'] =  valor 

2- fazendo a média
  valor = df.loc [df["col base"] == 'o quero', 'nome da coluna'].mean()
'''

classes = ["Iris-setosa", "Iris-versicolor", "Iris-virginica"] # coloco todas as classes em um lista

for i in classes:
  tabela.loc[tabela["classType"] == i, 'mean_sepalLength'] = tabela.loc[tabela['classType'] == i, 'sepalLength'].mean()


In [ ]:
tabela

sepalLength  sepalWidth  ...       classType  mean_sepalLength
0            5.1         3.5  ...     Iris-setosa             5.006
1            4.9         3.0  ...     Iris-setosa             5.006
2            4.7         3.2  ...     Iris-setosa             5.006
3            4.6         3.1  ...     Iris-setosa             5.006
4            5.0         3.6  ...     Iris-setosa             5.006
..           ...         ...  ...             ...               ...
145          6.7         3.0  ...  Iris-virginica             6.588
146          6.3         2.5  ...  Iris-virginica             6.588
147          6.5         3.0  ...  Iris-virginica             6.588
148          6.2         3.4  ...  Iris-virginica             6.588
149          5.9         3.0  ...  Iris-virginica             6.588

[150 rows x 6 columns]

e. Crie uma coluna de volume sabendo que o volume representa (pi x petallength x sepal_length^2)/3

In [ ]:
tabela['volume'] = ((3.14 * tabela["petalLength"] * (tabela["sepalLength"] ** 2)) / 3)

In [ ]:
tabela

sepalLength  sepalWidth  ...  mean_sepalLength      volume
0            5.1         3.5  ...             5.006   38.113320
1            4.9         3.0  ...             5.006   35.182653
2            4.7         3.2  ...             5.006   30.057127
3            4.6         3.1  ...             5.006   33.221200
4            5.0         3.6  ...             5.006   36.633333
..           ...         ...  ...               ...         ...
145          6.7         3.0  ...             6.588  244.321307
146          6.3         2.5  ...             6.588  207.711000
147          6.5         3.0  ...             6.588  229.952667
148          6.2         3.4  ...             6.588  217.262880
149          5.9         3.0  ...             6.588  185.815780

[150 rows x 7 columns]

f. Salve apenas o valor da classe, da média do sepal length e do volume desse dataset em um arquivo csv sem a coluna de index.

In [ ]:
novosDados = tabela.loc[:, ['classType','mean_sepalLength', 'volume']]
novosDados

classType  mean_sepalLength      volume
0       Iris-setosa             5.006   38.113320
1       Iris-setosa             5.006   35.182653
2       Iris-setosa             5.006   30.057127
3       Iris-setosa             5.006   33.221200
4       Iris-setosa             5.006   36.633333
..              ...               ...         ...
145  Iris-virginica             6.588  244.321307
146  Iris-virginica             6.588  207.711000
147  Iris-virginica             6.588  229.952667
148  Iris-virginica             6.588  217.262880
149  Iris-virginica             6.588  185.815780

[150 rows x 3 columns]

In [ ]:
novosDados.to_csv('saida_iris.csv', sep=';', index=False)